homeassistant/components/tellduslive/__init__.py

```diff
@@ -0,0 +1,289 @@
+            'type': 'TellStick ZNet Lite v2',
+            'ip': '31.208.248.64'
+        }]
+
+        async_track_time_interval(self._hass, self.update,
+                                  DEFAULT_UPDATE_INTERVAL)
+
+    async def update(self, *args):
+        """Periodically poll the servers for current state."""
+        _LOGGER.debug('Updating')
+        await self._sync()
+
+    async def _sync(self):
+        """Update local list of devices."""
+        if not self._client.update():
+            _LOGGER.warning('Failed request')
+
+        def identify_device(device):
+            """Find out what type of HA component to create."""
+            from tellduslive import (DIM, UP, TURNON)
+            if device.methods & DIM:
+                return 'light'
+            if device.methods & UP:
+                return 'cover'
+            if device.methods & TURNON:
+                return 'switch'
+            if device.methods == 0:
+                return 'binary_sensor'
+            _LOGGER.warning("Unidentified device type (methods: %d)",
+                            device.methods)
+            return 'switch'
+
+        async def discover(device_id, component):
+            """Discover the component."""
+            config_entries_key = '{}.{}'.format(component, 'mqtt')
+            async with self._hass.data[DATA_CONFIG_ENTRY_LOCK]:
+                if config_entries_key not in self._hass.data[
+                        CONFIG_ENTRY_IS_SETUP]:
+                    await self._hass.config_entries.async_forward_entry_setup(
+                        self._config_entry, component)
+                    self._hass.data[CONFIG_ENTRY_IS_SETUP].add(
+                        config_entries_key)
+
+            async_dispatcher_send(
+                self._hass, TELLDUS_DISCOVERY_NEW.format(component, DOMAIN),
+                device_id)
+
+        known_ids = {entity.device_id for entity in self.entities}
+        for device in self._client.devices:
+            if device.device_id in known_ids:
+                continue
+            if device.is_sensor:
+                for item in device.items:
+                    await discover((device.device_id, item.name, item.scale),
+                                   'sensor')
+            else:
+                await discover(device.device_id, identify_device(device))
+
+        for entity in self.entities:
+            entity.changed()
+
+    def device(self, device_id):
+        """Return device representation."""
+        return self._client.device(device_id)
+
+    def is_available(self, device_id):
+        """Return device availability."""
+        return device_id in self._client.device_ids
+
+
+class TelldusLiveEntity(Entity):
+    """Base class for all Telldus Live entities."""
+
+    def __init__(self, client, device_id):
+        """Initialize the entity."""
+        self._id = device_id
+        self._client = client
+        self._client.entities.append(self)
```

Comment:
This isn't allowed. We're not allowed to store entities in a shared container. Entities are a platform concern.

Reply:
Referencing my comment here too:
https://github.com/home-assistant/home-assistant/issues/17515#issuecomment-430442001

Reply:
Thank you Martin for reminding me of this. I've done the necessary refactoring now.

Will try to write the tox tests during the day but then I hope to be ready for merge.